src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/InputFlowControl.cs

```diff
@@ -25,7 +25,7 @@ internal sealed class InputFlowControl
     private FlowControl _flow;
```

Comment:
I think this type could be written using *Interlock.CompareExchange* pattern (described also inhttps://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.compareexchange?view=net-8.0). In my testing it would bring twice as better throughput with contention and 25% better perf without contention.

Reply:
Can you show the code you used and how you benchmarked it? I think the change you made to `DecrementActiveClientStreamCount` is fine, but you do have to be extra careful when changing previously locking code to be lock-free.

I imagine this involves looping in `FlowControl.TryUpdateWindow` if `FlowControl.Available` is changed from underneath you by another thread, since you cannot do the `bytes > maxUpdate` comparison and assign a new value to `Available` in a single interlocked operation, but even that's not sufficient.

It's also critical that both `InputFlowControl.TryAdvance` and `InputFlowControl.TryUpdateWindow` return false when `_flow.IsAborted` or else Kestrel might not return connection window space or double return connection window space respectively.

Reply:
I will prepare and showcase over the weekend including the benchmarks., In short I encoded the aborted flag and the available (`bool` and `uint`) in a `long` field with bit shifting and used interlocked over that single `long` field. I used two type of benchmarks, just "regular" ones and a benchmark very similar that @kouvel commented above.

Reply:
I have updated the PR with a commit including the possible changes for `InputFlowControl` type.

There are two types of performance tests I executed. One in part of the PR as a standard BDN tests, and one type using the following `Tester` class, which is based on the previous discussions of this PR.

```csharp
public class Tester()
{
    internal readonly InputFlowControl _flowControl = new(10000, 10);
    private const int Spin = 50;
    private int total = 0;
    private bool counterEnabled = false;

    internal async Task Run(InputFlowControl control)
    {
        for (int i = 0; i < 8; i++)
        {
            total = 0;
            bool abort = false;
            control.Reset();
            var t1 = Task.Factory.StartNew(() =>
            {
                int counter = 0;
                while (!abort)
                {
                    if (counterEnabled) counter++;
                    control.TryUpdateWindow(16, out _);
                }
                Interlocked.Add(ref total, counter);
            }, TaskCreationOptions.LongRunning);

            void Consumer()
            {
                int counter = 0;
                while (!abort)
                {
                    if (counterEnabled) counter++;
                    if (control.TryAdvance(1))
                    {
                        for (int j = 0; j < Spin; j++)
                        {
                        }
                    }
                }
                Interlocked.Add(ref total, counter);
            }

            var t2 = Task.Factory.StartNew(Consumer, TaskCreationOptions.LongRunning);
            var t3 = Task.Factory.StartNew(Consumer, TaskCreationOptions.LongRunning);
            var t4 = Task.Factory.StartNew(Consumer, TaskCreationOptions.LongRunning);

            Thread.Sleep(300);

            control.Reset();
            Interlocked.Exchange(ref counterEnabled, true);
            Thread.Sleep(1000);
            Interlocked.Exchange(ref abort, true);
            await Task.WhenAll(t1, t2, t3, t4);
            Console.WriteLine($"{control.GetType().Name}: {total},");
        }
    }
}
```
In the results below:
  - InputFlowControlState uses `Interlocked`.
  - InputFlowControl uses `lock`.
  - Spin 50 can be increased or reduced, it is a placeholder for any other work that might happen in between the locked operations (ie. sending data on the H2 streams).
  - In `Tester` class the higher the counter number, the more operations were performed.

```
InputFlowControl: 8921766,
InputFlowControl: 11849858,
InputFlowControl: 11833087,
InputFlowControl: 11729804,
InputFlowControl: 11639123,
InputFlowControl: 11758042,
InputFlowControl: 11570816,
InputFlowControl: 11501363,
InputFlowControlState: 18889513,
InputFlowControlState: 21042510,
InputFlowControlState: 21810964,
InputFlowControlState: 22845625,
InputFlowControlState: 22820625,
InputFlowControlState: 22374220,
InputFlowControlState: 22050317,
InputFlowControlState: 24744676,
```

Executing the benchmarks with BDN (part of the PR):

BEFORE:

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22631
12th Gen Intel Core i7-1255U, 1 CPU, 12 logical and 10 physical cores
.NET SDK=9.0.100-rc.2.24452.3
  [Host]     : .NET 10.0.0 (10.0.24.45517), X64 RyuJIT
  Job-EHBQOK : .NET 9.0.0 (9.0.24.42919), X64 RyuJIT

Server=True  Toolchain=.NET Core 9.0  InvocationCount=1
RunStrategy=Throughput  UnrollFactor=1

|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 38.26 ms | 0.814 ms | 2.386 ms | 26.14 |     - |     - |     - |      3 KB |
```

AFTER:

```
|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 17.73 ms | 1.028 ms | 3.030 ms | 56.40 |     - |     - |     - |      3 KB |
```

Reply:
Moved to https://github.com/dotnet/aspnetcore/pull/57968